src/components/SearchResultItem.tsx

```diff
@@ -8,23 +8,94 @@ interface ParentTextClickedFunc {
 }
 
 interface SearchResultItemProps {
-    data: LibraryUtilities.ItemData;
-    libraryContainer: LibraryContainer;
-    highlightedText: string;
-    pathToItem: LibraryUtilities.ItemData[];
-    onParentTextClicked: ParentTextClickedFunc;
-    detailed: boolean;
+    data: LibraryUtilities.ItemData,
+    libraryContainer: LibraryContainer,
+    highlightedText: string,
+    pathToItem: LibraryUtilities.ItemData[],
+    onParentTextClicked: ParentTextClickedFunc,
+    detailed: boolean,
+    index: number
 }
 
-interface SearchResultItemStates { }
+interface SearchResultItemStates {
+    selected: boolean
+}
 
 export class SearchResultItem extends React.Component<SearchResultItemProps, SearchResultItemStates> {
 
     constructor(props: SearchResultItemProps) {
         super(props);
+        this.state = ({ selected: false });
+    }
+
+    componentWillMount() {
+        window.addEventListener("keydown", this.handleKeyDown.bind(this));
+    }
+
+    componentWillUnmount() {
+        window.removeEventListener("keydown", this.handleKeyDown.bind(this));
+    }
+
+    // Update selection state and scroll current item into view if the selected item is not in view yet.
+    componentDidUpdate() {
+        this.updateSelectionState();
+
+        if (this.state.selected) {
+            let container = ReactDOM.findDOMNode(this.props.libraryContainer);
+            let currentItem = ReactDOM.findDOMNode(this);
+            let containerRect = container.getBoundingClientRect();
+            let currentRect = currentItem.getBoundingClientRect();
+
+            if (currentRect.top < currentRect.height) {
+                currentItem.scrollIntoView();
+            }
+
+            if (currentRect.bottom > containerRect.bottom) {
+                currentItem.scrollIntoView(false);
+            }
+        }
+    }
+
+    handleKeyDown(event: any) {
```

Comment:
Since every SearchResultItem is aware of its own state, this would have better performance than having libraryContainer in charge of the selection state of every item. For libraryContainer to do so, it has to rerender the whole 1000 items to refresh the selection state. However, if each SearchResultItem checks against selectionIndex to see whether itself is selected, only the two items that have state changed will be rerendered, other items would only respond to the event but not the rendering.